File: CurrencyConverter.cpp
#include <stdexcept>
#include <string>

#include <QLabel>
#include <QComboBox>
#include <QLineEdit>
#include <QVBoxLayout>
#include <QMessageBox>
#include <QJsonDocument>
#include <QJsonObject>

#include "CurrencyConverter.h"

double CurrencyConverter::parseExchangeRate(const std::string& jsonResponse, const std::string& targetCurrency) {
    QJsonDocument jsonDoc = QJsonDocument::fromJson(QByteArray::fromStdString(jsonResponse));
    if (jsonDoc.isNull() || !jsonDoc.isObject()) {
        throw std::runtime_error("Failed to parse JSON response");
    }

    QJsonObject jsonObj = jsonDoc.object();
    if (!jsonObj.contains("rates") || !jsonObj["rates"].isObject()) {
        throw std::runtime_error("Invalid JSON response: missing 'rates' object");
    }

    QJsonObject ratesObj = jsonObj["rates"].toObject();
    if (!ratesObj.contains(QString::fromStdString(targetCurrency))) {
        throw std::runtime_error("Target currency not found in rates");
    }

    QJsonValue rateValue = ratesObj[QString::fromStdString(targetCurrency)];
    if (!rateValue.isDouble()) {
        throw std::runtime_error("Invalid rate value for target currency");
    }

    return rateValue.toDouble();
}


double CurrencyConverter::getExchangeRate(const std::string& baseCurrency, const std::string& targetCurrency) {
    try {
        std::string url = "https://api.exchangerate-api.com/v4/latest/" + baseCurrency;
        std::string response = getHttpResponse(url);
        return parseExchangeRate(response, targetCurrency);
    }
    catch (const std::runtime_error& e) {
        QMessageBox::critical(nullptr, "Error", e.what());
        return -1;
    }
}

CurrencyConverter::CurrencyConverter(QWidget* parent) : QWidget(parent) {
    initUI();
}

void CurrencyConverter::initUI() {
    setWindowTitle("Currency Converter");

    baseCurrencyLabel = new QLabel("Base Currency:", this);
    targetCurrencyLabel = new QLabel("Target Currency:", this);
    amountLabel = new QLabel("Amount:", this);
    resultLabel = new QLabel("", this);

    baseCurrencyInput = new QComboBox(this);
    targetCurrencyInput = new QComboBox(this);
    QStringList currencies = { "USD", "EUR", "GBP", "PLN", "JPY", "AUD", "CAD", "CHF", "CNY", "SEK" };
    baseCurrencyInput->addItems(currencies);
    targetCurrencyInput->addItems(currencies);

    targetCurrencyInput->setCurrentText("PLN");

    amountInput = new QLineEdit(this);
    amountInput->setPlaceholderText("Enter amount");
    amountInput->setText("100");

    QVBoxLayout* layout = new QVBoxLayout;
    layout->addWidget(baseCurrencyLabel);
    layout->addWidget(baseCurrencyInput);
    layout->addWidget(targetCurrencyLabel);
    layout->addWidget(targetCurrencyInput);
    layout->addWidget(amountLabel);
    layout->addWidget(amountInput);
    layout->addWidget(resultLabel);

    setLayout(layout);

    connect(baseCurrencyInput, &QComboBox::activated, this, &CurrencyConverter::convertCurrency);
    connect(targetCurrencyInput, &QComboBox::activated, this, &CurrencyConverter::convertCurrency);
    connect(amountInput, &QLineEdit::returnPressed, this, &CurrencyConverter::convertCurrency);

    convertCurrency();
}

void CurrencyConverter::convertCurrency() {
    QString baseCurrency = baseCurrencyInput->currentText();
    QString targetCurrency = targetCurrencyInput->currentText();
    bool ok;
    double amount = amountInput->text().toDouble(&ok);

    if (!ok) {
        QMessageBox::warning(this, "Error", "Please enter a valid amount.");
        return;
    }

    double exchangeRate = getExchangeRate(baseCurrency.toStdString(), targetCurrency.toStdString());

    if (exchangeRate > 0) {
        double convertedAmount = amount * exchangeRate;
        resultLabel->setText(QString("%1 %2 is %3 %4")
            .arg(amount)
            .arg(baseCurrency)
            .arg(convertedAmount, 0, 'f', 2)
            .arg(targetCurrency));
    }
}



// #include "moc_CurrencyConverter.cpp"

File: CurrencyConverter.py
import sys
import requests

try:
    from PyQt6.QtWidgets import QApplication, QWidget, QLabel, QComboBox, QLineEdit, QVBoxLayout, QMessageBox
except ImportError:    
    from PyQt5.QtWidgets import QApplication, QWidget, QLabel, QComboBox, QLineEdit, QVBoxLayout, QMessageBox


class CurrencyConverter(QWidget):
    def __init__(self):
        super().__init__()
        self.initUI()

    def initUI(self):
        self.setWindowTitle('Konwerter Walut')

        # Labels
        self.base_currency_label = QLabel('Waluta bazowa:', self)
        self.target_currency_label = QLabel('Waluta docelowa:', self)
        self.amount_label = QLabel('Kwota:', self)
        self.result_label = QLabel('', self)

        # ComboBoxes
        self.base_currency_input = QComboBox(self)
        self.target_currency_input = QComboBox(self)
        currencies = ["USD", "EUR", "GBP", "PLN", "JPY", "AUD", "CAD", "CHF", "CNY", "SEK"]
        self.base_currency_input.addItems(currencies)
        self.target_currency_input.addItems(currencies)

        # Set default target currency to PLN
        self.target_currency_input.setCurrentText("PLN")

        # LineEdit
        self.amount_input = QLineEdit(self)
        self.amount_input.setPlaceholderText("Wprowadź kwotę")
        self.amount_input.setText("100")  # Set default amount to 100

        # Layout
        layout = QVBoxLayout()
        layout.addWidget(self.base_currency_label)
        layout.addWidget(self.base_currency_input)
        layout.addWidget(self.target_currency_label)
        layout.addWidget(self.target_currency_input)
        layout.addWidget(self.amount_label)
        layout.addWidget(self.amount_input)
        layout.addWidget(self.result_label)

        self.setLayout(layout)

        # Connect signals to convert_currency method
        self.base_currency_input.activated.connect(self.convert_currency)
        self.target_currency_input.activated.connect(self.convert_currency)
        self.amount_input.returnPressed.connect(self.convert_currency)

        # Perform initial conversion
        self.convert_currency()

    def get_exchange_rate(self, base_currency, target_currency):
        url = f"https://api.exchangerate-api.com/v4/latest/{base_currency}"
        try:
            response = requests.get(url)
            response.raise_for_status()  # Raise an error for bad responses
            data = response.json()

            if target_currency not in data['rates']:
                QMessageBox.critical(self, "Błąd", "Nie znaleziono kursu wymiany dla podanej waluty docelowej.")
                return None

            return data['rates'][target_currency]
        
        except requests.exceptions.RequestException as e:
            QMessageBox.critical(self, "Błąd", f"Błąd w połączeniu z API: {e}")
            return None

    def convert_currency(self):
        base_currency = self.base_currency_input.currentText()
        target_currency = self.target_currency_input.currentText()
        try:
            amount = float(self.amount_input.text())
        except ValueError:
            QMessageBox.warning(self, "Błąd", "Proszę wprowadzić prawidłową kwotę.")
            return
        
        exchange_rate = self.get_exchange_rate(base_currency, target_currency)
        
        if exchange_rate is not None:
            converted_amount = amount * exchange_rate
            self.result_label.setText(f"{amount} {base_currency} to {converted_amount:.2f} {target_currency}")

def main():
    app = QApplication(sys.argv)
    converter = CurrencyConverter()
    converter.show()
    sys.exit(app.exec())

if __name__ == '__main__':
    main()